promql/engine.go

```diff
@@ -1074,6 +1081,23 @@ func (ev *evaluator) matrixIterSlice(it *storage.BufferedSeriesIterator, mint, m
 		}
 	}
 
+	if maxt-mint > ev.interval && len(out) > 0 && out[len(out)-1].T >= mint {
```

Comment:
This is a layering violation, you're making presumptions both about what's in `out` and poking around `ev`. This function should be standalone, and any preconditions clearly stated.

Reply:
It's undocumented and not particularly elegant, but I wouldn't go as far as calling it a violation of any kind. Simply put, if any data in `out` overlaps with the requested `[mint, maxt]` range, then it's kept, and any data retrieved from `it` is appended to it (with duplicates dropped).

I'll think of a better way of expressing this (in code and documentation).